### DISCUSSION TITLE
zip in [@types/d3-array] should be more general

### DISCUSSION THREAD

**Discussion by einarpersson** [DefinitelyTyped/DefinitelyTyped#70425] [Issues with a @types package]
According to the [docs](https://d3js.org/d3-array/transform#zip) this should work (and it does work)
```
zip(["Alice", "Bob", "Carol"], [32, 13, 14]) // [["Alice", 32], ["Bob", 13], ["Carol", 14]]
```

However it gives the following type error ` Type 'number' is not assignable to type 'string'.`

```
export function zip<T>(...arrays: Array<ArrayLike<T>>): T[][];
```
is too restrictive it seems, as it requires all supplied arrays to have the same type of elements

[a comment by an automated account is not shown]

**Comment by Methuselah96** [comment]
Feel free to create a PR!

**Comment by einarpersson** [comment]
@Methuselah96 Tbh I am a bit overwhelmed by this repo and how to go about it. I might do it but here is my current workaround anyhow

In `./types/d3-array.d.ts`
```
import * as d3Array from 'd3-array';

declare module 'd3-array' {
  export function zip<T extends any[]>(...arrays: T): { [K in keyof T]: T[K] extends (infer U)[] ? U : never }[];
}
```
which seems to do the job.

And configure tsconfig with
```
"compilerOptions" : {
    "typeRoots": ["./types", "./node_modules/@types"]
}
```

Note: Since zip is really just a wrapper around transpose i suppose the transpose function also has the same problem.
